sql: put parallel stmts behind environment variable in v1.0

nvb opened the pull request:
Parallel statements were not safe for general use because of their
behavior when confronted with retryable errors. This was fixed in #17627,
but the fix is too big to cherry-pick back into v1.0.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/18002)
<!-- Reviewable:end -->

bdarnell commented:
Should we introduce this setting in master and then cherry-pick to 1.0? Otherwise, upgrading from 1.0.6 to 1.1 will effectively remove this setting (which shouldn't be too bad, but might be a little surprising). 

How safe is it to enable the setting? Should we just turn it off without offering a setting to enable it?

---

Review status: 0 of 4 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/18002#-:-KsjZ_e3dtFonESJGFU5:br5iq6s)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Can we use an env var instead and just set that to true in CI? That might
be enough and avoid problems.

On Tue, Aug 29, 2017, 16:00 Ben Darnell <notifications@github.com> wrote:

> Should we introduce this setting in master and then cherry-pick to 1.0?
> Otherwise, upgrading from 1.0.6 to 1.1 will effectively remove this setting
> (which shouldn't be too bad, but might be a little surprising).
>
> How safe is it to enable the setting? Should we just turn it off without
> offering a setting to enable it?
> ------------------------------
>
> Review status: 0 of 4 files reviewed at latest revision, all discussions
> resolved, some commit checks failed.
> ------------------------------
>
> *Comments from Reviewable
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/18002#-:-KsjZ_e3dtFonESJGFU5:br5iq6s>*
>
> —
> You are receiving this because your review was requested.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/18002#issuecomment-325782476>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AE135PgxChSkdhVVwZ_gdH1WE4MeTrGdks5sdG3RgaJpZM4PGbsZ>
> .
>
-- 

-- Tobias

nvanbenschoten commented:
I wanted to keep it behind some gate so that people playing around with parallel statements (eg. @robert-s-lee) could still have access in v1.0.6. Your point about the setting disappearing is a good one though. I think an env var is more appropriate.

nvb committed:
sql: put parallel stmts behind environment variable in v1.0

Parallel statements were not safe for general use because of their
behavior when confronted with retryable errors. This was fixed in #17627,
but the fix is too big to cherry-pick back into v1.0.

nvanbenschoten commented:
This is using an environment variable now. PTAL.

bdarnell commented:
Review status: 0 of 4 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/sql/parallel_stmts_test.go, line 456 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18002#-KtRv5VT0xh9Ok0cNzjR:-KtRv5VT0xh9Ok0cNzjS:b-rggiuy) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b00d4f0a8abe589277ffba818d4cc0a3a878f19/pkg/sql/parallel_stmts_test.go#L456)):*
> ```Go
> 
> 	expectErr := "parallel statement execution is disabled by default until v1.1"
> 	_, err := db.Exec(`INSERT INTO test.foo VALUES (1) RETURNING NOTHING`)
> ```

We shouldn't make `RETURNING NOTHING` an error, it should just fall back to serial execution. (`RETURNING NOTHING` is already not a guarantee of parallelism; you can have multiple `RETURNING NOTHING` statements on the same table and 1.0 wouldn't parallelize them but 1.1 can)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/18002)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
Review status: 0 of 4 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/sql/parallel_stmts_test.go, line 456 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18002#-KtRv5VT0xh9Ok0cNzjR:-KtRvZoqmCF3shp2oj-a:b-1vxfuw) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b00d4f0a8abe589277ffba818d4cc0a3a878f19/pkg/sql/parallel_stmts_test.go#L456)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

We shouldn't make `RETURNING NOTHING` an error, it should just fall back to serial execution. (`RETURNING NOTHING` is already not a guarantee of parallelism; you can have multiple `RETURNING NOTHING` statements on the same table and 1.0 wouldn't parallelize them but 1.1 can)
</blockquote></details>

So you think I should revert the environment variable I just added and instead just run all `RETURNING NOTHING` statements serially? That seems fine to me, but I'd hate for demos or benchmarks to suddenly run a lot slower. Since we're this close to 1.1 though, it's probably not a big issue.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/18002)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
I'd be fine with leaving the environment variable in to toggle between parallel and serial modes, it just shouldn't be an error to have `RETURNING NOTHING` (suddenly getting slower is better than suddenly failing)

nvanbenschoten commented:
Ok, that's an even better idea.
